src/test/socket/connect/test_connect.rs

```diff
@@ -0,0 +1,526 @@
+        fd: fd,
+        addr: Some(addr),
+        addr_len: std::mem::size_of_val(&addr) as u32,
+    };
+
+    let expected_errno = if sock_type == libc::SOCK_DGRAM {
+        None
+    } else if flag & libc::SOCK_NONBLOCK != 0 {
+        Some(libc::EINPROGRESS)
+    } else {
+        Some(libc::ECONNREFUSED)
+    };
+
+    test_utils::run_and_close_fds(&[fd], || check_connect_call(&args, expected_errno))
+}
+
+/// Test connect() to an address with port 0.
+fn test_port_zero(sock_type: libc::c_int, flag: libc::c_int) -> Result<(), String> {
+    let fd = unsafe { libc::socket(libc::AF_INET, sock_type | flag, 0) };
+    assert!(fd >= 0);
+
+    let addr = libc::sockaddr_in {
+        sin_family: libc::AF_INET as u16,
+        sin_port: 0u16.to_be(),
+        sin_addr: libc::in_addr {
+            s_addr: libc::INADDR_LOOPBACK.to_be(),
+        },
+        sin_zero: [0; 8],
+    };
+
+    let args = ConnectArguments {
+        fd: fd,
+        addr: Some(addr),
+        addr_len: std::mem::size_of_val(&addr) as u32,
+    };
+
+    let expected_errno = if sock_type == libc::SOCK_DGRAM {
+        None
+    } else if flag & libc::SOCK_NONBLOCK != 0 {
+        Some(libc::EINPROGRESS)
+    } else {
+        Some(libc::ECONNREFUSED)
+    };
+
+    test_utils::run_and_close_fds(&[fd], || check_connect_call(&args, expected_errno))
+}
+
+/// Test connect() after closing the socket.
+fn test_after_close(sock_type: libc::c_int, flag: libc::c_int) -> Result<(), String> {
+    let fd = unsafe { libc::socket(libc::AF_INET, sock_type | flag, 0) };
+    assert!(fd >= 0);
+
+    let rv = unsafe { libc::close(fd) };
+    assert_eq!(rv, 0);
+
+    let addr = libc::sockaddr_in {
+        sin_family: libc::AF_INET as u16,
+        sin_port: 11111u16.to_be(),
+        sin_addr: libc::in_addr {
+            s_addr: libc::INADDR_LOOPBACK.to_be(),
+        },
+        sin_zero: [0; 8],
+    };
+
+    let args = ConnectArguments {
+        fd: fd,
+        addr: Some(addr),
+        addr_len: std::mem::size_of_val(&addr) as u32,
+    };
+
+    check_connect_call(&args, Some(libc::EBADF))
+}
+
+/// Test connect() to a server listening on the given interface (optionally overriding
+/// the interface the client connects on).
+fn test_interface(
+    sock_type: libc::c_int,
+    flag: libc::c_int,
+    server_interface: libc::in_addr_t,
+    override_client_interface: Option<libc::in_addr_t>,
+) -> Result<(), String> {
+    let fd_server = unsafe { libc::socket(libc::AF_INET, sock_type | flag, 0) };
+    let fd_client = unsafe { libc::socket(libc::AF_INET, sock_type | flag, 0) };
+    assert!(fd_server >= 0);
+    assert!(fd_client >= 0);
+
+    // the server address
+    let mut server_addr = libc::sockaddr_in {
+        sin_family: libc::AF_INET as u16,
+        sin_port: 0u16.to_be(),
+        sin_addr: libc::in_addr {
+            s_addr: server_interface.to_be(),
+        },
+        sin_zero: [0; 8],
+    };
+
+    // bind on the server address
+    let rv = unsafe {
```

Comment:
Optional: in places like this the `assert` can be inlined, letting us get rid of the named variable. This is helpful to someone reading the code since they don't have to try to remember what `rv` is (not knowing whether or not it's used later). Unlike C/C++ `assert`s are never compiled out (only `debug_assert`s)

Reply:
I like to keep things that change state outside of assertions since I find them easy to miss when scanning through code (and rustfmt can make them confusing when they use several lines), but I definitely agree with not polluting the function with extra variables, especially these test functions that can get fairly long.

Instead I'm now using explicit scopes so that it's easier to see that `rv` won't be used again later:

```
{
    let rv = unsafe { libc::close(fd) };
    assert_eq!(rv, 0);
}
```